File: lib/decorators/draggable_container.dart
part of dock_spawn;

class DraggableContainer implements IDockContainer {
  String containerType;
  IDockContainer delegate;
  Element topLevelElement;
  Element dragHandle;
  Dialog dialog;
  DialogEventListener eventListener;

  DockManager get dockManager => delegate.dockManager;
  
  StreamSubscription<MouseEvent> mouseDownHandler;
  StreamSubscription<MouseEvent> mouseUpHandler;
  StreamSubscription<MouseEvent> mouseMoveHandler;
  
  Point2 dragOffset;
  Point2 previousMousePosition;
  
  DraggableContainer(this.dialog, this.delegate, this.topLevelElement, this.dragHandle) {
    containerType = delegate.containerType;
        
    mouseDownHandler = dragHandle.onMouseDown.listen(onMouseDown);
    topLevelElement.style.marginLeft = "${topLevelElement.$dom_offsetLeft}";
    topLevelElement.style.marginTop = "${topLevelElement.$dom_offsetTop}";
  }

  void destroy() {
    removeDecorator();
    delegate.destroy();
  }

  void saveState(Map<String, Object> state) {
    delegate.saveState(state);
  }
  
  void loadState(Map<String, Object> state) {
    delegate.loadState(state);
  }
  
  void setActiveChild(IDockContainer child) {
  }

  int get minimumAllowedChildNodes { return delegate.minimumAllowedChildNodes; }
  
  int get width {
    return delegate.width;
  }

  int get height {
    return delegate.height;
  }

  String get name {
    return delegate.name;
  }
  void set name(String value) {
    delegate.name = value;
  }
  
  void resize(int _width, int _height) {
    delegate.resize(_width, _height);
  }
  
  void performLayout(List<IDockContainer> children) {
    delegate.performLayout(children);
  }
  
  void removeDecorator() {
    mouseDownHandler.cancel();
  }
  
  Element get containerElement {
    return delegate.containerElement;
  }
  
  void onMouseDown(MouseEvent event) {
    _startDragging(event);
    previousMousePosition = new Point2(event.$dom_pageX, event.$dom_pageY);
    if (mouseMoveHandler != null) {
      mouseMoveHandler.cancel();
      mouseMoveHandler = null;
    }
    if (mouseUpHandler != null) {
      mouseUpHandler.cancel();
      mouseUpHandler = null;
    }
    
    mouseMoveHandler = window.onMouseMove.listen(onMouseMove);
    mouseUpHandler = window.onMouseUp.listen(onMouseUp);
  }
  
  void onMouseUp(MouseEvent event) {
    _stopDragging(event);
    mouseMoveHandler.cancel();
    mouseMoveHandler = null;
    mouseUpHandler.cancel();
    mouseUpHandler = null;
  }
  
  void _startDragging(MouseEvent event) {
    if (dialog.eventListener != null) {
      dialog.eventListener.onDialogDragStarted(dialog, event);
    }
    
    document.body.classes.add("disable-selection");
  }
  
  void _stopDragging(MouseEvent event) {
    if (dialog.eventListener != null) {
      dialog.eventListener.onDialogDragEnded(dialog, event);
    }
    document.body.classes.remove("disable-selection");
  }

  void onMouseMove(MouseEvent event) {
    Point2 currentMousePosition = new Point2(event.$dom_pageX, event.$dom_pageY);
    int dx = (currentMousePosition.x - previousMousePosition.x).toInt();
    int dy = (currentMousePosition.y - previousMousePosition.y).toInt();
    _performDrag(dx, dy);
    previousMousePosition = currentMousePosition; 
  }
  
  void _performDrag(int dx, int dy) {
    int left = dx + getPixels(topLevelElement.style.marginLeft);
    int top = dy + getPixels(topLevelElement.style.marginTop);
    topLevelElement.style.marginLeft = "${left}px";
    topLevelElement.style.marginTop = "${top}px";
  }
}

File: js/lib/decorators/DraggableContainer.js
dockspawn.DraggableContainer = function(dialog, delegate, topLevelElement, dragHandle)
{
    this.dialog = dialog;
    this.delegate = delegate;
    this.containerElement = delegate.containerElement;
    this.dockManager = delegate.dockManager;
    this.topLevelElement = topLevelElement;
    this.containerType = delegate.containerType;
    this.mouseDownHandler = new dockspawn.EventHandler(dragHandle, 'mousedown', this.onMouseDown.bind(this));
    this.topLevelElement.style.marginLeft = topLevelElement.offsetLeft + "px";
    this.topLevelElement.style.marginTop = topLevelElement.offsetTop + "px";
    this.minimumAllowedChildNodes = delegate.minimumAllowedChildNodes;
};

dockspawn.DraggableContainer.prototype.destroy = function()
{
    this.removeDecorator();
    this.delegate.destroy();
};

dockspawn.DraggableContainer.prototype.saveState = function(state)
{
    this.delegate.saveState(state);
};

dockspawn.DraggableContainer.prototype.loadState = function(state)
{
    this.delegate.loadState(state);
};

dockspawn.DraggableContainer.prototype.setActiveChild = function(child)
{
};

Object.defineProperty(dockspawn.DraggableContainer.prototype, "width", {
    get: function() { return this.delegate.width; }
});

Object.defineProperty(dockspawn.DraggableContainer.prototype, "height", {
    get: function() { return this.delegate.height; }
});

dockspawn.DraggableContainer.prototype.name = function(value)
{
    if (value)
        this.delegate.name = value;
    return this.delegate.name;
};

dockspawn.DraggableContainer.prototype.resize = function(width, height)
{
    this.delegate.resize(width, height);
};

dockspawn.DraggableContainer.prototype.performLayout = function(children)
{
    this.delegate.performLayout(children);
};

dockspawn.DraggableContainer.prototype.removeDecorator = function()
{
    if (this.mouseDownHandler)
    {
        this.mouseDownHandler.cancel();
        delete this.mouseDownHandler;
    }
};

dockspawn.DraggableContainer.prototype.onMouseDown = function(event)
{
    this._startDragging(event);
    this.previousMousePosition = { x: event.pageX, y: event.pageY };
    if (this.mouseMoveHandler)
    {
        this.mouseMoveHandler.cancel();
        delete this.mouseMoveHandler;
    }
    if (this.mouseUpHandler)
    {
        this.mouseUpHandler.cancel();
        delete this.mouseUpHandler;
    }

    this.mouseMoveHandler = new dockspawn.EventHandler(window, 'mousemove', this.onMouseMove.bind(this));
    this.mouseUpHandler = new dockspawn.EventHandler(window, 'mouseup', this.onMouseUp.bind(this));
};

dockspawn.DraggableContainer.prototype.onMouseUp = function(event)
{
    this._stopDragging(event);
    this.mouseMoveHandler.cancel();
    delete this.mouseMoveHandler;
    this.mouseUpHandler.cancel();
    delete this.mouseUpHandler;
};

dockspawn.DraggableContainer.prototype._startDragging = function(event)
{
    if (this.dialog.eventListener)
        this.dialog.eventListener.onDialogDragStarted(this.dialog, event);
    document.body.classList.add("disable-selection");
};

dockspawn.DraggableContainer.prototype._stopDragging = function(event)
{
    if (this.dialog.eventListener)
        this.dialog.eventListener.onDialogDragEnded(this.dialog, event);
    document.body.classList.remove("disable-selection");
};

dockspawn.DraggableContainer.prototype.onMouseMove = function(event)
{
    var currentMousePosition = new Point(event.pageX, event.pageY);
    var dx = Math.floor(currentMousePosition.x - this.previousMousePosition.x);
    var dy = Math.floor(currentMousePosition.y - this.previousMousePosition.y);
    this._performDrag(dx, dy);
    this.previousMousePosition = currentMousePosition;
};

dockspawn.DraggableContainer.prototype._performDrag = function(dx, dy)
{
    var left = dx + getPixels(this.topLevelElement.style.marginLeft);
    var top = dy + getPixels(this.topLevelElement.style.marginTop);
    this.topLevelElement.style.marginLeft = left + "px";
    this.topLevelElement.style.marginTop = top + "px";
};
